File: 680. Valid Palindrome II/validPalindrome.py
class Solution:
    def validPalindrome(self, s: str) -> bool:
        left = 0
        right = len(s) - 1
        if left >= right:
            return True
        while left < right:
            if s[left] != s[right]:
                return self.isPalindrome(s, left + 1, right) or self.isPalindrome(s, left, right - 1)
            left += 1
            right -= 1
        return True

    def isPalindrome(self, s, left, right):
        if left >= right:
            return True
        if s[left] == s[right]:
            return self.isPalindrome(s, left + 1, right - 1)
        else:
            return False

File: 680. Valid Palindrome II/validPalindrome.java
class Solution {
    public boolean validPalindrome(String s) {
        int left = 0;
        int right = s.length() - 1;
        if (left >= right) {
            return true;
        }
        while (left < right) {
            if (s.charAt(left) != s.charAt(right)) {
                return isPalindrome(s, left + 1, right) || isPalindrome(s, left, right - 1);
            }
            left ++;
            right --;
        }
        return true;
    }

    public boolean isPalindrome(String s, int left, int right) {
        if (left >= right) {
            return true;
        }
        if(s.charAt(left) == s.charAt(right)) {
            return isPalindrome(s, left + 1, right - 1);
        }
        return false;
    }
}